Simplify EclConfig

valentin-krasontovitsch opened the pull request:
**Issue**
Closes #3918


**Approach**
Confer [discussion in other PR](https://github.com/equinor/ert/pull/3887#discussion_r971671276)

## Pre review checklist

- [x] Added appropriate release note label
- [x] PR title captures the intent of the changes, and is fitting for release notes.
- [x] Commit history is consistent and clean, in line with the [contribution guidelines](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md).

Adding labels helps the maintainers when writing release notes. This is the [list of release note labels](https://github.com/equinor/ert/labels?q=release-notes).

valentin-krasontovitsch commented:
one of the github action test runs failed with 429 responses from github (too many requests) O_o

eivindjahren commented:
I believe this might introduce changed behavior as 0is used to indicate that NUM_CPU is read from somewhere else. I think we should first create some tests that fail if num_cpu is not handled correctly when not set in the config.

valentin-krasontovitsch commented:
based on the discussion we had i think we take care of all cases where `NUM_CPU=0` is used to signal something. do you agree that all occurences of the above are covered in the changes in this PR? 

tests sound like a good idea - what do you mean by "if num_cpu is not handled correctly when not set in the config" ? let's say num_cpu is not set in the config - what do we consider to be incorrect handling of that fact?

eivindjahren commented:
What I think would be reasonable behavior is:

* `res_config.num_cpu = None` when NUM_CPU is not set in the config.
* There is a function `get_num_cpu_from_ecl() -> Optional[int]` which returns the number of CPUs to use set in the eclipse configuration file (None if there is no such config or no such setting in that config)
`EnkfMain` then arbitrates between this sources with
```
def get_num_cpu(self):
  if self.resConfig().num_cpu is not None:
    return self.resConfig().num_cpu
  if self.get_num_cpu_from_ecl() is not None:
    return self.get_num_cpu_from_ecl()
  return 1
```

valentin-krasontovitsch commented:
well that is conceptionally definitely very nice : )

i can try to make that work!

I guess i'm still kind of timid, and want to change as little as possible / necessary, and don't think much about "what should this actually look like", so thanks for the nudge!

valentin-krasontovitsch commented:
just to make sure i understand - this would remove NUM_CPU from queue config, and move it up to ResConfig? Or does a call of `self.resConfig().num_cpu` actually under the hood look at queue config?

eivindjahren commented on src/clib/lib/enkf/enkf_state.cpp:
```diff
@@ -32,15 +32,13 @@ void enkf_state_initialize(rng_type *rng, enkf_fs_type *fs,
-    const bool fmt_file = ecl_config_get_formatted(ecl_config);
     char *header_file = ecl_util_alloc_exfilename(
-        run_path, eclbase, ECL_SUMMARY_HEADER_FILE, fmt_file, -1);
+        run_path, eclbase, ECL_SUMMARY_HEADER_FILE, DEFAULT_FORMATTED, -1);
```
So I would expect here that `ecl_config_get_formatted` and `ecl_config_type->io_config` would be deleted in this commit.

eivindjahren replied:
Ah, now I understand, the entire file is deleted later, nvm.

eivindjahren commented on src/clib/lib/enkf/ecl_config.cpp:
```diff
@@ -1,483 +0,0 @@
-        if (ui_return_get_status(ui_return) == UI_RETURN_OK)
-            ecl_config->have_eclbase = true;
-        else
-            util_abort("%s: failed to set eclbase format. Error:%s\n", __func__,
```
Abort is bad, but we should probably have similar error messages raised.

valentin-krasontovitsch replied:
eclbase is no longer used, so we can scrap the warning, too!

eivindjahren commented on src/clib/lib/enkf/ecl_config.cpp:
```diff
@@ -1,483 +0,0 @@
-    if (ui_return_get_status(ui_return) == UI_RETURN_OK)
-        ecl_config_set_data_file(ecl_config, data_file);
-    else
-        util_abort("%s: problem setting ECLIPSE data file (%s)\n", __func__,
```
Abort is bad, but we should probably have similar error messages raised.  It would be good to add a test that checks that good error messages are given (ie. `pytest.raises(ValueError, match="{good message}"):`)

valentin-krasontovitsch replied:
validate data file just checks for its existence, so also here i will reference the validation of files, which is one of the problems that we took and put all the way over there: #3985

eivindjahren commented on src/clib/lib/enkf/ecl_config.cpp:
```diff
@@ -1,483 +0,0 @@
-    if (ui_return_get_status(ui_return) == UI_RETURN_OK)
-        ecl_config_set_grid(ecl_config, grid_file);
-    else
-        util_abort("%s: failed to set grid file:%s  Error:%s \n", __func__,
```
Abort is bad, but we should probably have similar error messages raised. I see that you check whether the file has the correct type. It would be good to add a test that checks that good error messages are given (ie. `pytest.raises(ValueError, match="{good message}"):`)

eivindjahren commented on src/clib/lib/enkf/ecl_config.cpp:
```diff
@@ -1,483 +0,0 @@
-        config_content_get_value_as_abspath(config, REFCASE_KEY);
-
-    if (!ecl_config_load_refcase(ecl_config, refcase_path))
-        fprintf(stderr, "** Warning: loading refcase:%s failed \n",
```
We should probably have similar warning message printed, and test that we get the expected warning: `pytest.warns(UserWarning, match="..."):`.

valentin-krasontovitsch replied:
gonna check out the python function now used for refcase loading for its error handling and see what can be done there, as per other comment

eivindjahren commented on src/clib/lib/enkf/ecl_config.cpp:
```diff
@@ -1,483 +0,0 @@
-
-    if (ecl_config->can_restart)
-        fprintf(
-            stderr,
```
We should probably have similar warning message printed, and test that we get the expected warning: pytest.warns(UserWarning, match="..."):.

eivindjahren commented on src/clib/lib/enkf/ecl_config.cpp:
```diff
@@ -1,483 +0,0 @@
-    case ECL_LAB_UNITS:
-        return "CM";
-    default:
-        util_abort("%s: unit system enum value:%d not recognized \n", __func__,
```
Abort is bad, but we should probably have similar error messages raised. It might be that this error message is already good, but we should add a test for such a message (ie. pytest.raises(ValueError, match="{good message}"):)

valentin-krasontovitsch replied:
code no longer used

eivindjahren commented on src/clib/lib/enkf/ecl_config.cpp:
```diff
@@ -1,483 +0,0 @@
-    case ECL_LAB_UNITS:
-        return "ATMA";
-    default:
-        util_abort("%s: unit system enum value:%d not recognized \n", __func__,
```
Abort is bad, but we should probably have similar error messages raised. It might be that this error message is already good, but we should add a test for such a message (ie. pytest.raises(ValueError, match="{good message}"):)

oysteoh commented on tests/libres_tests/res/enkf/test_ecl_config.py:
```diff
@@ -1,93 +1,59 @@
-    assert dfile == ec.getDataFile()
+    ec = EclConfig(data_file=dfile)
+    assert dfile == ec._data_file
+    assert 1 == ec.num_cpu
```
Use 'ec.num_cpu == 1' instead of '1 == ec.num_cpu' (Yoda-conditions)

valentin-krasontovitsch replied:
ah but isn't it assert expected is equal to actual? or is it always exactly the other way around? :thinking:

valentin-krasontovitsch replied:
anyways it is fixed now. the real truth will remain a mystery of the cosmos to me, as i seemingly cannot remember either way.

valentin-krasontovitsch commented:
@eivindjahren there's only three things left that might need to raise warnings or exceptions. it's gonna be difficult to insert error checking for those, as they basically just call python or in one case a c method from the ecl repo. with this simplification, we are at the mercy of those functions doing something meaningful.

also just got an idea, i could wrap it all in a try and except thingy, hoping that the functions raise something and then we can work with that. i'm gonna check out a bit how the functions are written with respect to error handling.

if you have any good ideas about this, please let me know!

regarding validation of like is a refcase file, grid file, data file given, i skipped that, as i'm not sure what we want to have, and where to put it - cf. also #3985

oysteoh commented on src/ert/_c_wrappers/enkf/ecl_config.py:
```diff
@@ -1,207 +1,96 @@
-        return self._get_last_history_restart()
+        return cls(**ecl_config_args)
+
+    def _load_grid(grid_file: str) -> Optional[EclGrid]:
```
i might not understand this fully, but the assignment of `self._grid_file` in the initialization and then the later `EclConfig._load_grid(grid_file)` .... Why do we need to add the `grid_file` as a parameter? Could the `_load_grid(..)` be called without an argument and inside the function just use the already assigned `self._grid_file`?

valentin-krasontovitsch replied:
ja indeed my intention here is that i had in the back of my head that there should be as little logic in the subconfig objects as possible, or at least all parsing should happen in the res config, or something along those lines.

so i wanted to make it clear that this function can basically stand on its own, I guess it doesn't even need to be defined as part of the class, can just be used by it. i'm gonna try to make it so.

valentin-krasontovitsch replied:
that sounds strangely reasonable, and like it doesn't even need deep understanding of object oriented stuff :sweat_smile:

ok so simple instance methods, that don't get the files as arguments, but grab them from the instance, got it!

oysteoh commented on src/ert/_c_wrappers/enkf/ecl_config.py:
```diff
@@ -1,207 +1,96 @@
+        self.grid = EclConfig._load_grid(grid_file)
+
+        self._refcase_file = refcase_file
+        self.refcase = EclConfig._load_refcase(refcase_file)
```
Argument 1 to "_load_refcase" of "EclConfig" has incompatible type "Optional[str]"; expected "str"

valentin-krasontovitsch replied:
also fixed!

oysteoh commented on src/ert/_c_wrappers/enkf/ecl_config.py:
```diff
@@ -1,207 +1,96 @@
-        return self._get_last_history_restart()
+        return cls(**ecl_config_args)
+
+    def _load_grid(grid_file: str) -> Optional[EclGrid]:
```
`def _load_grid(self, grid_file: str) -> Optional[EclGrid]`

valentin-krasontovitsch replied:
sorry, that should have been a class method (to make it clear that it can be moved outta there)

valentin-krasontovitsch replied:
never mind, making it a stand alone function, see above

oysteoh commented on src/ert/_c_wrappers/enkf/ecl_config.py:
```diff
@@ -1,207 +1,96 @@
+            raise ValueError(f"grid file {grid_file} does not have expected type")
+        return EclGrid.load_from_file(grid_file)
+
+    def _load_refcase(refcase_file: str) -> EclSum:
```
`self` 😄

valentin-krasontovitsch replied:
indeed : )

valentin-krasontovitsch replied:
gonna take same approach as above

valentin-krasontovitsch commented on src/ert/_c_wrappers/enkf/ecl_config.py:
```diff
@@ -1,207 +1,96 @@
+    ):
+        self._data_file = data_file
+        self._grid_file = grid_file
+        self.grid = EclConfig._load_grid(grid_file)
```
i should really check these annotations myself :sweat_smile: 
gonna check out what went wrong here!

valentin-krasontovitsch replied:
yes, yes, of course, i see it now - the file can be None, that's silly of me. it is fixed now!

valentin-krasontovitsch commented:
so checking out loading the refcase, which used the `__init__` function of the `EclSum` class found in the `ecl` repo, there are exceptions raised in python. no technical details (as in strings explaining anything) are passed up from the C code called under the hood in `ecl`, but we get a message a la `Failed to create summary instance from argument:[INSERT FILENAME]` of type `IOError`

i am welcome to hear suggestions on how we can improve on this.

in the meantime, i'm gonna try to write a test or two, to tickle out these error messages, as suggested by Eivind!

valentin-krasontovitsch commented:
oh, two questions:

- do we want to catch the errors and make them warnings? as this happens at the moment during init of config, we might want to say "whatever if you don't manage to load this refile, just warn about it, and move in with life i.e. execution of code"? or do we want to fail with an exception? i guess for the former we should also have a plan about what happens downstream in the program, i.e. make sure that we catch such a situtation (no refcase where one is expected).
- where do i put a test of this? it's not really config parsing, it's testing logic in the config object. do i create a new folder `config` in `tests/ert_tests`?

valentin-krasontovitsch commented:
okay i think i addressed everything - the things i fixed or think we don't need to fix i resolved, and the only thing left to discuss i think is where to test and how to handle undesired events when loading stuff (see comment above).

I've pushed the (non trivial / minute) changes in extra commits, for your convenience, so you can easily review them.

Please also tell me how you feel about the handling of config content as a dict (and the simplification needed there to extract values that are packed as listlists) - see own commit d903f1c4e55b7fc2d69db968d1227f1a6a63aa36

eivindjahren commented on tests/test_config_parsing/test_ecl_config.py:
```diff
@@ -17,4 +17,59 @@ def test_ecl_config_dict_creates_equal_config(config_dict):
+_|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|
+___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|__
+_|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|
+___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|___|__
```
He he, funny but perhaps just say 4 lines of this? Also, maybe a clear indication that this is not what a grid file looks like for those not initiated into dark magic.

valentin-krasontovitsch replied:
pytest.raises, gotcha - i'm learning every day! and will add a comment : )

valentin-krasontovitsch replied:
done!

eivindjahren commented on tests/test_config_parsing/test_ecl_config.py:
```diff
@@ -17,4 +17,59 @@ def test_ecl_config_dict_creates_equal_config(config_dict):
+    with open(grid_file, "w+", encoding="utf-8") as grid_file_handler:
+        grid_file_handler.write(grid_file_content)
+    try:
+        EclConfig(grid_file=grid_file)
```
You can use

```python
with pytest.raises(ValueError, match=grid_file):
  _ = EclConfig(grid_file=grid_file)
```

match can also contain other parts of the error message.

valentin-krasontovitsch replied:
will do, thanks : )

valentin-krasontovitsch replied:
done!

eivindjahren commented on tests/test_config_parsing/test_ecl_config.py:
```diff
@@ -17,4 +17,59 @@ def test_ecl_config_dict_creates_equal_config(config_dict):
+    with open(refcase_file, "w+", encoding="utf-8") as refcase_file_handler:
+        refcase_file_handler.write(refcase_file_content)
+    try:
+        EclConfig(refcase_file=refcase_file)
```
Also here we can use
```python
with pytest.raises(ValueError, match=refcase_file):
  _ = EclConfig(refase_file=refcase_file)
```

match can also contain other parts that should be in the error message.

valentin-krasontovitsch replied:
sure thing!

valentin-krasontovitsch replied:
done!

eivindjahren commented on src/ert/_c_wrappers/enkf/ecl_config.py:
```diff
@@ -1,207 +1,91 @@
+            EclFileEnum.ECL_GRID_FILE,
+            EclFileEnum.ECL_EGRID_FILE,
+        ]
+        if not get_file_type(self._grid_file) in ecl_grid_file_types:
```
:robot:

valentin-krasontovitsch replied:
not sure about this one - my hunch is, it should `X not in LIST` rather thatn `not X in LIST`?

valentin-krasontovitsch replied:
if that's it, then it's done

valentin-krasontovitsch replied:
that was it!

eivindjahren reviewed (changes requested):
Just some small comments here, I think the implementation is ready, but it would be nice to use pytest.raises in the tests.

oysteoh reviewed (commented):
i think from my part it looks allright now ! but i'll leave it up to @eivindjahren  to approve the changes he have requested 👍

eivindjahren reviewed (approved):
Just needs to fix style and all should be good :rocket:

dafeda commented on src/clib/lib/enkf/enkf_state.cpp:
```diff
@@ -302,6 +300,8 @@ static std::pair<fw_load_status, std::string> enkf_state_internalize_results(
 
     enkf_fs_type *sim_fs = run_arg_get_sim_fs(run_arg);
     int last_report = time_map_get_last_step(enkf_fs_get_time_map(sim_fs));
+    // TODO investigate usage of last history: is last_report < 0 possible? is
```
I think it's better to create an issue than adding TODO's to code as they are quite hidden.

valentin-krasontovitsch replied:
ARGH thank you for catching that! that's a personal note that should never have been merged :grimacing:

valentin-krasontovitsch replied:
gonna make a PR that removes this immediately.

valentin-krasontovitsch replied:
fixed in https://github.com/equinor/ert/pull/4037
